pkg/internalapis/featurestates/featurestates.go

```diff
@@ -0,0 +1,418 @@
+	for range ticker.C {
+		func() {
+			pendingCRUpdatesObj.lock.Lock()
+			defer pendingCRUpdatesObj.lock.Unlock()
+
+			ctx, log := logger.GetNewContextWithLogger()
+			for namespace, updateRequired := range pendingCRUpdatesObj.namespaceUpdateMap {
+				if !updateRequired {
+					continue
+				}
+				log.Infof("feature state update is required for namespace: %q", namespace)
+				// check if CR is present on the namespace
+				featurestateCR := &featurestatesv1alpha1.CnsCsiSvFeatureStates{}
+				err := controllerRuntimeClient.Get(ctx, client.ObjectKey{Name: SVFeatureStateCRName,
+					Namespace: namespace}, featurestateCR)
+				if err == nil {
+					// Attempt to Update CR
+					featurestateCR.Spec.FeatureStates = pendingCRUpdatesObj.latestFeatureStates
+					err = controllerRuntimeClient.Update(ctx, featurestateCR)
+					if err != nil {
+						log.Errorf("failed to update cnsCsiSvFeatureStates CR instance in the namespace: %q, Err: %v", namespace, err)
+						continue
+					}
+					pendingCRUpdatesObj.namespaceUpdateMap[namespace] = false
+					log.Infof("updated cnsCsiSvFeatureStates CR instance in the namespace: %q", namespace)
+				} else {
+					if apierrors.IsNotFound(err) {
+						log.Infof("cnsCsiSvFeatureStates CR instance is not present in the namespace: %q. "+
+							"Creating CR with latest feature switch state, Err: %v", namespace, err)
+						// attempt to Create the CR
+						cnsCsiSvFeatureStates := &featurestatesv1alpha1.CnsCsiSvFeatureStates{
+							ObjectMeta: metav1.ObjectMeta{Name: SVFeatureStateCRName, Namespace: namespace},
+							Spec: featurestatesv1alpha1.CnsCsiSvFeatureStatesSpec{
+								FeatureStates: pendingCRUpdatesObj.latestFeatureStates,
+							},
+						}
+						err = controllerRuntimeClient.Create(ctx, cnsCsiSvFeatureStates)
+						if err != nil {
+							log.Errorf("failed to create cnsCsiSvFeatureStates CR instance in the "+
+								"namespace: %q, Err: %v", namespace, err)
+							continue
+						}
+						pendingCRUpdatesObj.namespaceUpdateMap[namespace] = false
+						log.Infof("created cnsCsiSvFeatureStates CR instance in the namespace: %q", namespace)
+					} else {
+						log.Errorf("failed to check if cnsCsiSvFeatureStates CR is present in the namespace :%q, err: %v", namespace, err)
+					}
+				}
+			}
+		}()
+	}
+}
+
+// enqueueFeatureStateUpdatesForAllWorkloadNamespaces helps enqueue featurestates updates for all workload namespaces
+func (pendingCRUpdatesObj *pendingCRUpdates) enqueueFeatureStateUpdatesForAllWorkloadNamespaces(ctx context.Context, featurestates []featurestatesv1alpha1.FeatureState) {
+	pendingCRUpdatesObj.lock.Lock()
+	defer pendingCRUpdatesObj.lock.Unlock()
+
+	log := logger.GetLogger(ctx)
+	pendingCRUpdatesObj.latestFeatureStates = featurestates
+	for namespace := range pendingCRUpdatesObj.namespaceUpdateMap {
+		pendingCRUpdatesObj.namespaceUpdateMap[namespace] = true
+		log.Infof("enqueued CR updates for workload namespace: %q", namespace)
```

Comment:
Wouldn't this be too much of logging when the supervisor cluster has hundreds of namespaces? Can we move it to debug and maybe have an info log before the for-loop?

Reply:
adding an info log after the for loop

```golang
for namespace := range pendingCRUpdatesObj.namespaceUpdateMap {
    pendingCRUpdatesObj.namespaceUpdateMap[namespace] = true
}
log.Infof("enqueued CR updates for all workload namespaces")
```